Publisher refactor

adambabik opened the pull request:
The goal is to remove all dependencies on status-go/status-react specific stuff like signals from publisher package. All that specific code should be encapsulated in `services/shhext` exclusively.

adambabik committed:
rename publisher.Service to publisher.Publisher

adambabik committed:
move status-go specific logic out from publisher

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (21)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 64ea6bc2 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1508/1/) | 2019-06-28 07:08:54 | ~54 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190628-070759-64ea6b.zip) |
| :heavy_check_mark: | 64ea6bc2 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1508/1/) | 2019-06-28 07:12:46 | ~4 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190628-070758-64ea6b.zip) |
| :heavy_check_mark: | 64ea6bc2 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1508/1/) | 2019-06-28 07:14:20 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190628-070757-64ea6b.aar) |
| | | | | | | |
| :heavy_check_mark: | f262b1b2 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1508/2/) | 2019-06-28 07:33:02 | ~40 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190628-073221-f262b1.zip) |
| :heavy_check_mark: | f262b1b2 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1508/2/) | 2019-06-28 07:34:48 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190628-073221-f262b1.zip) |
| :heavy_check_mark: | f262b1b2 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1508/2/) | 2019-06-28 07:37:42 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190628-073220-f262b1.aar) |
| | | | | | | |
| :heavy_check_mark: | be12384c | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1508/3/) | 2019-06-28 08:34:34 | ~32 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190628-083402-be1238.zip) |
| :heavy_check_mark: | be12384c | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1508/3/) | 2019-06-28 08:36:33 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190628-083400-be1238.zip) |
| :heavy_check_mark: | be12384c | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1508/3/) | 2019-06-28 08:39:21 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190628-083359-be1238.aar) |
| | | | | | | |
| :heavy_check_mark: | b94872c9 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1508/4/) | 2019-06-28 13:07:24 | ~37 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190628-130646-b94872.zip) |
| :heavy_check_mark: | b94872c9 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1508/4/) | 2019-06-28 13:09:24 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190628-130646-b94872.zip) |
| :heavy_check_mark: | b94872c9 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1508/4/) | 2019-06-28 13:12:12 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190628-130645-b94872.aar) |
| | | | | | | |
| :heavy_check_mark: | 5bfcdbe9 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1508/5/) | 2019-06-29 07:53:49 | ~43 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190629-075306-5bfcdb.zip) |
| :heavy_check_mark: | 5bfcdbe9 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1508/5/) | 2019-06-29 07:55:50 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190629-075305-5bfcdb.zip) |
| :heavy_check_mark: | 5bfcdbe9 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1508/5/) | 2019-06-29 07:58:24 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190629-075304-5bfcdb.aar) |
| | | | | | | |
| :heavy_check_mark: | 3dc5f3de | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1508/6/) | 2019-06-29 07:57:28 | ~27 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190629-075700-3dc5f3.zip) |
| :heavy_check_mark: | 3dc5f3de | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1508/6/) | 2019-06-29 07:59:39 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190629-075659-3dc5f3.zip) |
| :heavy_check_mark: | 3dc5f3de | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1508/6/) | 2019-06-29 08:03:32 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190629-075825-3dc5f3.aar) |
| | | | | | | |
| :heavy_check_mark: | ef1c59e1 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1508/7/) | 2019-07-01 06:42:06 | ~45 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190701-064121-ef1c59.zip) |
| :heavy_check_mark: | ef1c59e1 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1508/7/) | 2019-07-01 06:44:00 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190701-064120-ef1c59.zip) |
| :heavy_check_mark: | ef1c59e1 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1508/7/) | 2019-07-01 06:48:57 | ~7 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190701-064119-ef1c59.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 9300f9a1 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1508/8/) | 2019-07-01 06:45:53 | ~42 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190701-064510-9300f9.zip) |
| :heavy_check_mark: | 9300f9a1 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1508/8/) | 2019-07-01 06:47:38 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190701-064509-9300f9.zip) |
| :heavy_check_mark: | 9300f9a1 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1508/8/) | 2019-07-01 06:55:38 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190701-064858-9300f9.aar) |
| | | | | | | |
| :heavy_check_mark: | 04b7047f | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1508/9/) | 2019-07-01 09:11:50 | ~42 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190701-091108-04b704.zip) |
| :heavy_check_mark: | 04b7047f | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1508/9/) | 2019-07-01 09:13:48 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190701-091107-04b704.zip) |
| :heavy_check_mark: | 04b7047f | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1508/9/) | 2019-07-01 09:16:11 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190701-091106-04b704.aar) |

adambabik commented on messaging/publisher/publisher_test.go:
```diff
@@ -164,12 +151,7 @@ func (s *ServiceTestSuite) TestTopic() {
-
-	err = s.alice.ProcessMessage(dedupMessage2)
+	err = s.alice.ProcessMessage(message2, []byte("1"))
 	s.Require().NoError(err)
```
In this line, I am getting:
```
Error:      	Received unexpected error:
                 no payload
```
@cammellos, any idea why is it so? It's a bit strange because it's a public message and `newMessage2, err := s.bob.sendContactCode()` succeeds.

cammellos replied:
That's correct, basically in those messages there's no payload, as we don't transfer any data other than the actual bundles, we should probably not consider it an error as it's a perfectly valid message

adambabik replied:
Ah ok that makes sense. We indeed do not send any message just contact code. The only thing is that it worked before my changes... :D

adambabik committed:
fix incorrectly refactored parts

adambabik committed:
handle errors properly in ProcessMessage

adambabik committed:
fix test when sending coontact code without payload

adambabik commented on messaging/publisher/publisher.go:
```diff
@@ -0,0 +1,455 @@
+	case nil:
+		msg.Payload = response
+	case chat.ErrDeviceNotFound:
+		// TODO(adam): why is this condition needed? What's the different if it it's false?
```
@cammellos can you please help me with this one? And two lines below, I don't understand what it means if public keys are different in this case.

cammellos replied:
`ErrDeviceNotFound` is thrown when we receive a message that is not targeted to our device.

Our own devices are manage explicitly, meaning that if we see a new device with our own public key (say A1 & A2), we will not be targeting `A2` by default, it needs to be explicitly enabled by the user (by calling `EnableInstallation(deviceID)`. That's for security reason, as we expect the user to be in control of his own devices, and he knows which one to trust (so if a user steals our own key, we won't be disclosing our own data to the new device).
That means that if we received a message from our own public key, not for our device, it's because the user *explicitly* did not want our device to receive it (because it's not a paired device).

When it comes to devices we don't own (say B1 and B2), those are managed implicitly, as there's no source of truth, so, A1 will send to B1 & B2 even if they are not paired, as we don't really know which one to trust.
So if we receive a message from B that is not targeting our device, we want to let B know that we are also to be targeted (which is what this signal does, upon receiving a message that is not targeted to us, we will reply with our bundle, so we are included in the next message).

Not sure if it's clear

cammellos replied:
Also, this signal will not need to be sent to the client (any client) anymore, eventually this can all be handled in the backend (`status-go` or whatever), basically upon seeing that we are not targeted, we just reply with a message, and we send maximum a message every x hours (we don't want to reply to any message, otherwise it's too much traffic)

adambabik replied:
Ok, so for the case where the public keys (from the message and our) are the same, first case, we can return `ErrNotPairedDevice` which probably gonna be only logged and skipped.

For the second case, maybe `ErrInvalidDeviceTarget` can be returned and in such a case, the signal will be sent.

Does it sound legitimate? I would even put this logic into `HandleMessage` to return a well specified error.

cammellos replied:
yes, sounds much better!

adambabik replied:
Actually, the second case is covered:
```go
        msg := msgs[s.config.InstallationID]
	if msg == nil {
		msg = msgs[noInstallationID]
	}

	// We should not be sending a signal if it's coming from us, as we receive our own messages
	if msg == nil && *theirIdentityKey != myIdentityKey.PublicKey {
		return nil, ErrDeviceNotFound
	}
```

I also realised that the first option is not gonna happen because as you said we don't send a message to not paired device. However, if some other implementation did that, we will have a panic because the code above is followed by `payload := msg.GetPayload()` and `msg` being `nil`.

I will cover only this case with `ErrNotPairedDevice `.

cammellos replied:
> I also realised that the first option is not gonna happen because as you said we don't send a message to not paired device

It actually does happen because of the way whisper works, essentially we use the whisper pk to target messages, so anyone with your pk will receive the message:
```
{
  installation-id-1: encrypted-payload1
  installation-id-2: encrypted-payload2
}
```

So any of your installations will receive this, but only `installation-1` and `installation-2` will be able to decrypt the content, while `installation-3` will receive the message, but will have no content that can be decrypted

adambabik replied:
Ah right, I did it correctly (https://github.com/status-im/status-go/pull/1508/commits/5bfcdbe9c8a325eac3bb65614df15c7144a0f786#diff-e155aad09e72a37bdf6f38432d5b9ab6R241) but described it in a wrong way. If a message is sent from device 1 or 2, of course device 3 will receive it but the installation won't match. Without the change mentioned above it could panic or this case is handled somewhere else earlier?

cammellos replied:
Interesting, it definitely does not panic as we would have noticed, but it's not handled, so just by looking at the code it should panic.

would it be possible that this always return true? ` *theirIdentityKey != myIdentityKey.PublicKey` is using underneath `BigInt` and possibly is comparing pointers instead (as I understand it should use `Cmp`). Status react also makes the same check so we never noticed.

adambabik replied:
Ah, that might be a case. This is never true... In the console client, we check if explicitly with `Cmp` to avoid that: https://github.com/status-im/status-console-client/blob/6009b73feb651d0df793f758c1e630b99af4b051/protocol/client/stream.go#L81. I will fix it.

adambabik committed:
handle a case when a device is not paired

adambabik committed:
change log level for chat.ErrNotPairedDevice

cammellos reviewed (approved):
looks ok to me, a bit hard to see what has changed in publisher (maybe we could use `git mv`? )
but as I understand the main changes are:

1) Initialization of the db has been moved out of publisher into shhextservice
2) Changed the inteface of process message so that it does not take a Dedup struct, but rather two args
3) Signals moved away from publisher

Anything else? (apart from various refactoring)
As I understand this is not a breaking change for status-react, correct?

adambabik commented:
@cammellos good point, I will try use `git mv` but generally nothing changed. I just removed `Init*` methods and added one to set dependencies.

Your summary is correct. There are no other changes except adding `chat.ErrNotPairedDevice`.

Yes, this should be backward compatible from status-react point of view.

adambabik committed:
bump VERSION to 0.29.0-beta.1

adambabik committed:
fix comparing public keys

adambabik committed:
fix formatting
